File: src/entities/Unit.js
// Базовый класс юнита
export class Unit {
    constructor(scene, gridX, gridY, isEnemy = false) {
        this.scene = scene;
        this.gridX = gridX;
        this.gridY = gridY;
        this.isEnemy = isEnemy;
        this.isDead = false;
        
        // Базовые характеристики
        this.maxHp = 30;
        this.hp = this.maxHp;
        this.damage = 10;
        this.attackSpeed = 1.0; // секунды между атаками
        this.range = 2;
        this.size = { width: 1, height: 1 };
        this.color = 0xFFFFFF;
        
        // Визуальные элементы
        this.sprite = null;
        this.hpBar = null;
        
        // Таймеры атак
        this.lastAttackTime = 0;
        
        this.createVisuals();
    }

    createVisuals() {
        const worldPos = this.scene.gridSystem.getWorldPosition(this.gridX, this.gridY);
        
        // Создаем спрайт юнита
        this.sprite = this.scene.add.rectangle(
            worldPos.x, 
            worldPos.y, 
            this.size.width * this.scene.gridSystem.cellSize * 0.8,
            this.size.height * this.scene.gridSystem.cellSize * 0.8,
            this.color
        );
        
        // Добавляем рамку
        this.scene.add.rectangle(
            worldPos.x, 
            worldPos.y, 
            this.size.width * this.scene.gridSystem.cellSize * 0.8,
            this.size.height * this.scene.gridSystem.cellSize * 0.8,
            0x000000,
            0
        ).setStrokeStyle(2, 0x333333);
        
        // Создаем полоску здоровья
        this.createHpBar(worldPos.x, worldPos.y);
        
        // Если это враг, делаем его серым
        if (this.isEnemy) {
            this.sprite.setFillStyle(0x666666);
        }
    }

    createHpBar(x, y) {
        const barWidth = this.size.width * this.scene.gridSystem.cellSize * 0.6;
        const barHeight = 4;
        const barY = y - (this.size.height * this.scene.gridSystem.cellSize * 0.4);
        
        // Фон полоски здоровья
        this.hpBarBg = this.scene.add.rectangle(x, barY, barWidth, barHeight, 0x333333);
        
        // Сама полоска здоровья
        this.hpBar = this.scene.add.rectangle(x, barY, barWidth, barHeight, 0x00FF00);
        
        this.updateHpBar();
    }

    updateHpBar() {
        if (!this.hpBar) return;
        
        const barWidth = this.size.width * this.scene.gridSystem.cellSize * 0.6;
        const hpPercent = this.hp / this.maxHp;
        const currentWidth = barWidth * hpPercent;
        
        this.hpBar.setSize(currentWidth, 4);
        
        // Меняем цвет в зависимости от здоровья
        if (hpPercent > 0.6) {
            this.hpBar.setFillStyle(0x00FF00); // зеленый
        } else if (hpPercent > 0.3) {
            this.hpBar.setFillStyle(0xFFFF00); // желтый
        } else {
            this.hpBar.setFillStyle(0xFF0000); // красный
        }
    }

    setPosition(x, y) {
        if (this.sprite) {
            this.sprite.setPosition(x, y);
            if (this.hpBar) {
                this.hpBar.setPosition(x, y - (this.size.height * this.scene.gridSystem.cellSize * 0.4));
            }
            if (this.hpBarBg) {
                this.hpBarBg.setPosition(x, y - (this.size.height * this.scene.gridSystem.cellSize * 0.4));
            }
        }
    }

    getSize() {
        return this.size;
    }

    getPosition() {
        return { x: this.gridX, y: this.gridY };
    }

    canAttack() {
        const currentTime = this.scene.time.now;
        return (currentTime - this.lastAttackTime) >= (this.attackSpeed * 1000);
    }

    attack(target) {
        if (!this.canAttack() || this.isDead) return false;
        
        this.lastAttackTime = this.scene.time.now;
        
        // Наносим урон
        target.takeDamage(this.damage);
        
        // Создаем визуальный эффект атаки
        this.createAttackEffect(target);
        
        console.log(`${this.constructor.name} атакует ${target.constructor.name} на ${this.damage} урона`);
        
        return true;
    }

    createAttackEffect(target) {
        const startPos = this.sprite;
        const endPos = target.sprite;
        
        // Создаем линию атаки
        const graphics = this.scene.add.graphics();
        graphics.lineStyle(3, 0xFFD700, 0.8);
        graphics.lineBetween(startPos.x, startPos.y, endPos.x, endPos.y);
        
        // Удаляем эффект через 200мс
        this.scene.time.delayedCall(200, () => {
            graphics.destroy();
        });
    }

    takeDamage(damage) {
        this.hp = Math.max(0, this.hp - damage);
        this.updateHpBar();
        
        if (this.hp <= 0) {
            this.die();
        }
        
        // Эффект получения урона
        this.createDamageEffect();
    }

    createDamageEffect() {
        // Мигание красным
        this.sprite.setTint(0xFF0000);
        this.scene.time.delayedCall(100, () => {
            this.sprite.clearTint();
        });
    }

    die() {
        this.isDead = true;
        
        // Анимация смерти
        this.scene.tweens.add({
            targets: this.sprite,
            alpha: 0,
            scaleX: 0,
            scaleY: 0,
            duration: 300,
            onComplete: () => {
                this.destroy();
            }
        });
        
        // Удаляем полоску здоровья
        if (this.hpBar) this.hpBar.destroy();
        if (this.hpBarBg) this.hpBarBg.destroy();
        
        console.log(`${this.constructor.name} погиб!`);
    }

    destroy() {
        if (this.sprite) this.sprite.destroy();
        if (this.hpBar) this.hpBar.destroy();
        if (this.hpBarBg) this.hpBarBg.destroy();
    }

    isAlive() {
        return !this.isDead && this.hp > 0;
    }

    getStats() {
        return {
            hp: this.hp,
            maxHp: this.maxHp,
            damage: this.damage,
            attackSpeed: this.attackSpeed,
            range: this.range
        };
    }
}

File: assets/Scripts/Units/Unit.cs
using UnityEngine;
using UnityEngine.UI;
using System.Collections;
using MergePrototype.Core;
using MergePrototype.Systems;

namespace MergePrototype.Units
{
    public abstract class Unit : MonoBehaviour
    {
        [Header("Unit Stats")]
        public int maxHp = 30;
        public int hp;
        public int damage = 10;
        public float attackSpeed = 1.0f;
        public int range = 2;
        public Vector2Int size = new Vector2Int(1, 1);
        public Color unitColor = Color.white;
        
        [Header("Position")]
        public int gridX;
        public int gridY;
        public bool isEnemy = false;
        
        [Header("Visual Components")]
        public SpriteRenderer spriteRenderer;
        public GameObject hpBarObject;
        public Image hpBarFill;
        public Image hpBarBackground;
        
        // State
        protected bool isDead = false;
        protected float lastAttackTime = 0f;
        protected int stars = 0; // For merge system
        
        // References
        protected GridSystem gridSystem;
        
        protected virtual void Awake()
        {
            // Initialize components
            spriteRenderer = GetComponent<SpriteRenderer>();
            if (spriteRenderer == null)
            {
                spriteRenderer = gameObject.AddComponent<SpriteRenderer>();
            }
            
            // Initialize HP
            hp = maxHp;
            
            // Find grid system
            gridSystem = FindObjectOfType<GridSystem>();
        }
        
        protected virtual void Start()
        {
            CreateVisuals();
            CreateHpBar();
        }
        
        protected virtual void Update()
        {
            if (!isDead && !isEnemy)
            {
                // Auto-attack logic for player units
                TryAttack();
            }
        }
        
        protected virtual void CreateVisuals()
        {
            // Set sprite color
            spriteRenderer.color = isEnemy ? Color.gray : unitColor;
            
            // Set scale based on size
            Vector3 scale = new Vector3(
                size.x * GameConfig.CELL_SIZE * 0.8f,
                size.y * GameConfig.CELL_SIZE * 0.8f,
                1f
            );
            transform.localScale = scale;
            
            // Create a simple colored square sprite
            CreateColoredSprite();
        }
        
        private void CreateColoredSprite()
        {
            // Create a simple white texture
            Texture2D texture = new Texture2D(1, 1);
            texture.SetPixel(0, 0, Color.white);
            texture.Apply();
            
            Sprite sprite = Sprite.Create(texture, new Rect(0, 0, 1, 1), new Vector2(0.5f, 0.5f));
            spriteRenderer.sprite = sprite;
        }
        
        protected virtual void CreateHpBar()
        {
            // Create HP bar canvas
            GameObject canvasObject = new GameObject("HPBarCanvas");
            canvasObject.transform.SetParent(transform);
            canvasObject.transform.localPosition = Vector3.zero;
            
            Canvas canvas = canvasObject.AddComponent<Canvas>();
            canvas.renderMode = RenderMode.WorldSpace;
            canvas.sortingOrder = 10;
            
            CanvasScaler scaler = canvasObject.AddComponent<CanvasScaler>();
            scaler.dynamicPixelsPerUnit = 10;
            
            // Create HP bar background
            GameObject bgObject = new GameObject("HPBarBackground");
            bgObject.transform.SetParent(canvasObject.transform);
            bgObject.transform.localPosition = new Vector3(0, -0.4f, 0);
            
            hpBarBackground = bgObject.AddComponent<Image>();
            hpBarBackground.color = new Color(0.2f, 0.2f, 0.2f, 0.8f);
            
            RectTransform bgRect = hpBarBackground.rectTransform;
            bgRect.sizeDelta = new Vector2(size.x * GameConfig.CELL_SIZE * 0.6f, 4f);
            bgRect.anchorMin = new Vector2(0.5f, 0.5f);
            bgRect.anchorMax = new Vector2(0.5f, 0.5f);
            
            // Create HP bar fill
            GameObject fillObject = new GameObject("HPBarFill");
            fillObject.transform.SetParent(bgObject.transform);
            fillObject.transform.localPosition = Vector3.zero;
            
            hpBarFill = fillObject.AddComponent<Image>();
            hpBarFill.color = Color.green;
            
            RectTransform fillRect = hpBarFill.rectTransform;
            fillRect.sizeDelta = new Vector2(size.x * GameConfig.CELL_SIZE * 0.6f, 4f);
            fillRect.anchorMin = new Vector2(0, 0.5f);
            fillRect.anchorMax = new Vector2(0, 0.5f);
            fillRect.anchoredPosition = new Vector2(-fillRect.sizeDelta.x / 2, 0);
            
            hpBarObject = canvasObject;
            UpdateHpBar();
        }
        
        protected virtual void UpdateHpBar()
        {
            if (hpBarFill == null) return;
            
            float hpPercent = (float)hp / maxHp;
            hpBarFill.fillAmount = hpPercent;
            
            // Change color based on health
            if (hpPercent > 0.6f)
            {
                hpBarFill.color = Color.green;
            }
            else if (hpPercent > 0.3f)
            {
                hpBarFill.color = Color.yellow;
            }
            else
            {
                hpBarFill.color = Color.red;
            }
        }
        
        public virtual void SetPosition(float x, float y)
        {
            transform.position = new Vector3(x, y, 0);
        }
        
        public virtual void SetGridPosition(int x, int y)
        {
            gridX = x;
            gridY = y;
        }
        
        public virtual Vector2Int GetSize()
        {
            return size;
        }
        
        public virtual bool IsAlive()
        {
            return !isDead && hp > 0;
        }
        
        public virtual bool IsEnemy()
        {
            return isEnemy;
        }
        
        public virtual void TakeDamage(int damageAmount)
        {
            if (isDead) return;
            
            hp -= damageAmount;
            hp = Mathf.Max(0, hp);
            
            UpdateHpBar();
            
            if (hp <= 0)
            {
                Die();
            }
        }
        
        protected virtual void Die()
        {
            isDead = true;
            
            // Play death animation
            StartCoroutine(DeathAnimation());
            
            Debug.Log($"{GetType().Name} died!");
        }
        
        protected virtual IEnumerator DeathAnimation()
        {
            // Fade out animation
            float duration = 0.3f;
            float elapsed = 0f;
            Color originalColor = spriteRenderer.color;
            
            while (elapsed < duration)
            {
                elapsed += Time.deltaTime;
                float alpha = Mathf.Lerp(1f, 0f, elapsed / duration);
                spriteRenderer.color = new Color(originalColor.r, originalColor.g, originalColor.b, alpha);
                yield return null;
            }
            
            // Destroy HP bar
            if (hpBarObject != null)
            {
                Destroy(hpBarObject);
            }
            
            // Remove from grid
            if (gridSystem != null)
            {
                gridSystem.RemoveUnit(this);
            }
            
            // Destroy game object
            Destroy(gameObject);
        }
        
        protected virtual void TryAttack()
        {
            if (Time.time - lastAttackTime < attackSpeed) return;
            
            // Find targets
            var targets = FindTargets();
            if (targets.Count > 0)
            {
                Attack(targets);
                lastAttackTime = Time.time;
            }
        }
        
        protected virtual System.Collections.Generic.List<Unit> FindTargets()
        {
            var targets = new System.Collections.Generic.List<Unit>();
            
            if (gridSystem == null) return targets;
            
            // Get all enemy units
            var enemyUnits = isEnemy ? gridSystem.GetPlayerUnits() : gridSystem.GetEnemyUnits();
            
            // Find units in range
            foreach (var unit in enemyUnits)
            {
                if (unit == null || !unit.IsAlive()) continue;
                
                float distance = Vector2.Distance(
                    new Vector2(gridX, gridY),
                    new Vector2(unit.gridX, unit.gridY)
                );
                
                if (distance <= range)
                {
                    targets.Add(unit);
                }
            }
            
            return targets;
        }
        
        protected virtual void Attack(System.Collections.Generic.List<Unit> targets)
        {
            // Override in derived classes for specific attack logic
            foreach (var target in targets)
            {
                if (target != null && target.IsAlive())
                {
                    target.TakeDamage(damage);
                }
            }
        }
        
        public virtual bool Merge(Unit other)
        {
            // Check if units can merge (same type and adjacent)
            if (GetType() != other.GetType()) return false;
            if (isEnemy != other.isEnemy) return false;
            
            // Check if adjacent
            float distance = Vector2.Distance(
                new Vector2(gridX, gridY),
                new Vector2(other.gridX, other.gridY)
            );
            
            if (distance > 1.5f) return false;
            
            // Merge units
            stars++;
            maxHp = Mathf.RoundToInt(maxHp * 1.5f);
            hp = maxHp;
            damage = Mathf.RoundToInt(damage * 1.3f);
            attackSpeed = Mathf.Max(0.1f, attackSpeed * 0.9f);
            
            UpdateHpBar();
            
            // Destroy the other unit
            other.Die();
            
            Debug.Log($"{GetType().Name} merged! Stars: {stars}");
            return true;
        }
        
        public virtual UnitStats GetStats()
        {
            return new UnitStats
            {
                hp = hp,
                maxHp = maxHp,
                damage = damage,
                attackSpeed = attackSpeed,
                range = range,
                stars = stars
            };
        }
        
        public int GetStars()
        {
            return stars;
        }
        
        public int GetSellPrice()
        {
            var unitData = GameConfig.GetUnitData(GetUnitType());
            return unitData.sellPrice + (stars * unitData.sellPricePerStar);
        }
        
        protected abstract GameConfig.UnitType GetUnitType();
    }
    
    [System.Serializable]
    public struct UnitStats
    {
        public int hp;
        public int maxHp;
        public int damage;
        public float attackSpeed;
        public int range;
        public int stars;
    }
}
